earth_enterprise/src/common/FileAccessor.cpp

```diff
@@ -0,0 +1,64 @@
+#include "FileAccessor.h"
+
+std::unique_ptr<AbstractFileAccessor> AbstractFileAccessor::getAccessor(const std::string &fname) {
+  return std::unique_ptr<POSIXFileAccessor>(new POSIXFileAccessor());
+}
+
+// ****************************************************************************
+// ***  POSIXFileAccessor
+// ****************************************************************************
+int POSIXFileAccessor::Open(const std::string &fname, const char *mode, int flags, mode_t createMask) {
+  int result;
+  if (mode) {
+    if (strcmp(mode, "w") == 0) {
+      result = khOpenForWrite(fname, createMask);
+    }
+    else {
+      result = khOpenForRead(fname);
+    }
+  }
+  else {
+    result = khOpen(fname, flags, createMask);
+  }
+  fileDescriptor = result;
+  return result;
+}
+
+int POSIXFileAccessor::FsyncAndClose() {
+  int result = khFsyncAndClose(fileDescriptor);
+  this->invalidate();
+  return result;
+}
+
+int POSIXFileAccessor::Close() {
+  int result = khClose(fileDescriptor);
+  this->invalidate();
+  return result;
+}
+
+bool POSIXFileAccessor::PreadAll(void* buffer, size_t size, off64_t offset) {
+  return khPreadAll(fileDescriptor, buffer, size, offset);
+}
+
+bool POSIXFileAccessor::PwriteAll(const void* buffer, size_t size, off64_t offset) {
+  return khPwriteAll(fileDescriptor, buffer, size, offset);
+}
+
+bool POSIXFileAccessor::ReadStringFromFile(const std::string &filename, std::string &str, std::uint64_t limit) {
+  return khReadStringFromFile(filename, str, limit);
+}
+
+bool POSIXFileAccessor::Exists(const std::string &filename) {
+  return khExists(filename);
+}
+
+int POSIXFileAccessor::fgets(void *buf, size_t bufsize) {
+  return read(fileDescriptor, buf, bufsize);
```

Comment:
I don't think this is going to do exactly the same thing as `fgets`. `fgets` stops reading if it encounters a newline character. `read` will just read `bufsize` bytes, stopping short only if it hits the end of the file.

Reply:
I can't find a similar function that works on file descriptors. Your best bet may be to read the entire file into a string, create a file pointer from the stream, and then run fgets on that pointer. The downside is that we'll be holding the whole file in memory, but from my investigation this file is typically only a few lines. You've already got ReadStringFromFile to read the file into a string, and `fmemopen` will create a file pointer from the string. Then you can use `fgets` on the string as if it was a file.

Reply:
The only place right now that uses this `fgets` method is reading the whole file and processing each line. What if we get rid of the `fgets` method here, and just replace it with a method that reads the entire contents of a file and returns a vector of strings containing all the lines?